Stroop Test Enhancements

eschramm opened the pull request:
For an upcoming project, the following parameters to the Stroop test has been requested:
- text should always be black when not incongruent with the color displayed
- it should be possible to mix the "box" format with the "text" format inside a single Stroop test

This necessitated an addition to the `ORKStroopResult` for the mixed-style test.

eschramm committed:
Stroop enhancements

- add option for always black text and simply randomizing word colors
- add ability to randomize between box and text format
- add box/text to ORKStroopResult

eschramm committed:
add image assets for Stroop enhancements

eschramm committed:
bugfix - timer will fire after step complete otherwise

eschramm committed:
remove unnecessary header import

mmertsock commented on ResearchKit/ActiveTasks/ORKStroopContentView.m:
```diff
@@ -169,32 +170,33 @@ - (void)setUseGridLayoutForButtons:(bool)useGridLayoutForButtons{
 - (UIColor *)colorLabelColor {
-    return _colorLabel.textColor;
+    if ([_colorLabel.textColor isEqual:[UIColor blackColor]]) {
+        return _colorLabel.text;
```
`_colorLabel.text` seems incorrect

eschramm replied:
Ah, this is no longer needed and was from before a refactor in handling the state for filing data for the ORKStroopResult

mmertsock commented on ResearchKit/ActiveTasks/ORKStroopContentView.m:
```diff
@@ -40,6 +40,7 @@ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 UIStackViewAlignment alignment = UILayoutConstraintAxisHorizontal;
 CGFloat labelHeight = 250.0;
 CGFloat labelWidth = 250.0;
+NSArray <NSLayoutConstraint *> *boxConstraints;
```
Why is this a global instead of a field member of ORKStroopContentView?

eschramm replied:
Good catch

mmertsock commented on ResearchKit/ActiveTasks/ORKStroopResult.h:
```diff
@@ -66,6 +66,11 @@ ORK_CLASS_AVAILABLE
+/**
+ The Stroop style displayed - "box" or "text"
+ */
+@property (nonatomic, copy, nullable) NSString *stroopStyle;
```
Is there a reason for this to be NSString instead of ORKStroopStyle?

eschramm replied:
I was debating this - since this should only be a box or text. Have updated things to use the ORKStroopStyle enum now.

mmertsock replied:
Oh, I see how this works. It's not the style of the overall Stroop test, but the style of an individual prompt within the test. So your original design using a string may make more sense than repurposing the ORKStroopStyle enum.

eschramm committed:
fix warnings - type should be UILayoutConstraintAxis

eschramm committed:
PR fixes

- move boxConstraints to have object internal scope
- remove obsolete colorLabelColor parameter/getter function
- change ORKStroopResult.stroopStyle to return ORKStroopStyle
- update naming of ORKStroopStyle to better match use in ORKStroopResult

mmertsock commented on ResearchKit/ActiveTasks/ORKStroopContentView.m:
```diff
@@ -37,16 +37,16 @@ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 @implementation ORKStroopContentView {
     UILabel *_colorLabel;
     UIStackView *_buttonStackView;
+    NSArray <NSLayoutConstraint *> *boxConstraints;
```
Should start with underline to match objc/ResearchKit naming conventions

eschramm committed:
fix boxConstraints naming to _boxConstraints

eschramm committed:
revert ORKStroopResult.stroopStyle to String

overcyn commented on ResearchKit/ActiveTasks/ORKStroopStepViewController.m:
```diff
@@ -233,13 +236,14 @@ - (void)start {
     stroopResult.startTime = _startTime;
     stroopResult.endTime =  [NSProcessInfo processInfo].systemUptime;
-    stroopResult.color = color;
+    stroopResult.color = _correctColorString;
```
Shouldn't this value be the displayed color rather than the correct color? i.e. If the stroopStyle=ORKStroopStyleTextAlwaysBlack then ORKStroopResult.color="Black"?

eschramm replied:
Yeah, I wasn't so sure about how to handle - but I agree, it's a more correct description of the state displayed.

overcyn commented on ResearchKit/Common/ORKTypes.h:
```diff
@@ -371,6 +371,35 @@ ORK_EXTERN ORKTrailMakingTypeIdentifier const ORKTrailMakingTypeIdentifierA;
+    /**
+     Display large text with a color word, text color depends upon randomizeVisualAndColorAlignment setting.
+     */
+    ORKStroopStyleText,
```
What do you think about something like this for the names? ORKStroopStyleBox, ORKStroopStyleBlackText, ORKStroopStyleColoredText, ORKStroopStyleRandomizeColoredTextOrBox.

eschramm replied:
That does seem more clear

eschramm commented:
I would like to merge Friday, if possible. Let me know if there are any issues.

overcyn commented on ResearchKit/ActiveTasks/ORKStroopStepViewController.m:
```diff
@@ -253,31 +257,62 @@ - (void)startNextQuestionOrFinish {
     }
+    _singleTestStroopStyle = isText ? @"text" : @"box";
     [self setButtonsEnabled];
     _startTime = [NSProcessInfo processInfo].systemUptime;
```
This is the logic table for StroopStyle and RandomizeVisualAndColorAlignment. Is this correct? ORKStroopStyleTextAlwaysBlack will sometimes show non-black text. Also this seems very complex.

StroopStyle | Randomize | Behavior
-|-|-
Box|False| random RGBY box
Box|True| random RGBY box
Text|False| random RGBY color with matching RGBY text
Text|True| random RGBY color with random RGBY text
TextAlwaysBlack|False| black color with random RGBY text
TextAlwaysBlack|True| (black color with random RGBY text) or (random RGBY color with non-matching random RGBY text)
RandomizeTextOrBox|False| (random RGBY box) or (black color with random RGBY text)
RandomizeTextOrBox|True| (random RGBY box) or (black color with random RGBY text) or (random RGBY color with non-matching random RGBY text)

eschramm replied:
I was initially trying to keep the code minimally different than what came from the stable branch. But I agree that this is hard to read now with the multiple options. Have refactored into a clearer flow now.

overcyn replied:
I like the changes. LGTM

eschramm committed:
improve enum naming options for ORKStroopStyle

eschramm committed:
PR fixes

- refactor logic to be more succinct
- set color to match text color when black
